plugin/src/main/java/org/jenkinsci/plugins/workflow/cps/CpsFlowExecution.java

```diff
@@ -632,14 +634,71 @@ private CpsScript parseScript() throws IOException {
 
             for (Entry<String, String> e : loadedScripts.entrySet()) {
                 shell.reparse(e.getKey(), e.getValue());
             }
-        } catch (RuntimeException | Error x) {
+        } catch (Exception x) {
+            // Suspected groovyjarjarasm.asm.MethodTooLargeException or a
+            // org.codehaus.groovy.control.MultipleCompilationErrorsException
+            // whose collection of errors refers to MethodTooLargeException.
+            // Per review comments, we do not want to statically compile a
+            // dependency on the groovyjarjarasm.asm.MethodTooLargeException
+            // internals, so gauge hitting it via String name comparisons.
+            // Other cases may be (subclasses of) RuntimeException or Error.
+            Exception mtlEx = null;
+            int ecCount = 0;
+
+            // Clean up first
             closeShells();
-            throw x;
+
+            if (x.getClass().getSimpleName().equals("MethodTooLargeException")) {
+                mtlEx = x;
+                ecCount = 1;
+            } else if (x instanceof MultipleCompilationErrorsException) {
+                ErrorCollector ec = ((MultipleCompilationErrorsException) x).getErrorCollector();
+                ecCount = ec.getErrorCount();
+
+                for (int i = 0; i < ecCount; i++) {
+                    Exception ex = ec.getException(i);
+                    if (ex == null)
+                        continue;
+
+                    LOGGER.log(Level.FINE, "Collected Exception #" + i + ": " + ex.toString());
+                    if (ex.getClass().getSimpleName().equals("MethodTooLargeException")) {
+                        mtlEx = ex;
+                        break;
+                    }
+                }
+            }
+
+            if (mtlEx == null || ecCount < 1) {
+                // Some other exception type, or collection did not include MTL, rethrow as-is
+                throw x;
+            }
+
+            String msg = "FAILED to parse WorkflowScript (the pipeline script) due to MethodTooLargeException";
+            if (ecCount > 1) {
+                msg += " (and other issues)";
+            }
+            // Short message suffices, not much that a pipeline developer
+            // can do with the stack trace into the guts of groovy
+            msg += "; please refactor to simplify code structure and/or move logic to a Jenkins Shared Library: ";
+            msg += mtlEx.getMessage();
+
+            if (ecCount > 1) {
+                // Not squashing with explicit MethodTooLargeException
+                // re-thrown below, in this codepath we have other errors.
+                throw new RuntimeException(msg, x);
+            } else {
+                // Do not confuse pipeline devs by a wall of text in the
+                // build console, but let the full context be found in
+                // server log with some dedication.
+                LOGGER.log(Level.FINE, mtlEx.getMessage());
```

Comment:
If we dropped that "severe" but short logging of just the prepared message into server console log that was just above (via earlier review bit => commit 89bbd4036f93a6ba8d3b226f6503c2cd6cad5046), should this detailed but "fine"-level log also disappear? Originally it was the detailed context for those couple of lines, that could be seen with a boost of verbosity, now it is probably just noise. (Or that dropped message could be revived at a less visible verbosity, e.g. make these "FINE" and "FINEST"?..)

Reply:
After some consideration (and per comment just above here), I've restored that server log message - but at a "FINER" level so it is normally no-op. Still, can be an aid to Jenkins admins (e.g. developers who would reproduce the problem on their local instance).